test/discr/discr_ops_test.py

```diff
@@ -0,0 +1,287 @@
+
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscreteGradient(discr_space)
+
+    for ndim in range(1, 6):
+        # DiscreteLp Vector
+        vsize = 3
+        intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+        space = FunctionSpace(intvl)
+        discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+        dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+        # Gradient
+        grad = DiscreteGradient(discr_space)
+        grad(dom_vec)
+
+    # DiscreteLp Vector
+    ndim = 2
+    vsize = 3
+    intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+    space = FunctionSpace(intvl)
+    discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+    dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+    # Gradient
+    grad = DiscreteGradient(discr_space)
+    grad_vec = grad(dom_vec)
+    assert len(grad_vec) is ndim
+
+    # Adjoint operator
+    ran_vec = grad.range.element(
+        [ndvolume(vsize, ndim) ** 2] * ndim)
+    adj = grad.adjoint
+    adj_vec = adj(ran_vec)
+    lhs = ran_vec.inner(grad_vec)
+    rhs = dom_vec.inner(adj_vec)
+    assert not lhs == 0
+    assert not rhs == 0
+    assert lhs == rhs
+
+
+def test_discrete_divergence():
+    """Discretized spatial divergence operator."""
+
+    # Invalid arguments
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscreteDivergence(discr_space)
+
+    # Check result of operator with explicit summation
+    data = np.array([[0., 1., 2., 3., 4.],
+                     [1., 2., 3., 4., 5.],
+                     [2., 3., 4., 5., 6.]])
+
+    # DiscreteLp
+    space = FunctionSpace(Rectangle([0, 0], [6, 2.5]))
+    discr_space = uniform_discr_fromspace(space, data.shape)
+
+    # Operator instance
+    div = DiscreteDivergence(discr_space)
+
+    # Apply operator
+    dom_vec = div.domain.element([data, data])
+    div_dom_vec = div(dom_vec)
+
+    # computation of divergence with helper function
+    dx0, dx1 = discr_space.grid.stride
+    df0 = finite_diff(data, axis=0, dx=dx0, zero_padding=True, edge_order=2)
+    df1 = finite_diff(data, axis=1, dx=dx1, zero_padding=True, edge_order=2)
+
+    assert all_equal(df0 + df1, div_dom_vec.asarray())
+
+    # Adjoint operator
+    adj_div = div.adjoint
+    ran_vec = div.range.element(data ** 2)
+    adj_div_ran_vec = adj_div(ran_vec)
+
+    # Adjoint condition
+    lhs = ran_vec.inner(div_dom_vec)
+    rhs = dom_vec.inner(adj_div_ran_vec)
+    assert not lhs == 0
+    assert not rhs == 0
+    assert almost_equal(lhs, rhs)
+
+    # Higher dimensional arrays
+    for ndim in range(1, 6):
+
+        # DiscreteLp Vector
+        lin_size = 3
+        space = FunctionSpace(IntervalProd([0.] * ndim, [lin_size] * ndim))
+        discr_space = uniform_discr_fromspace(space, [lin_size] * ndim)
+
+        # Divergence
+        div = DiscreteDivergence(discr_space)
+        dom_vec = div.domain.element([ndvolume(lin_size, ndim)] * ndim)
+        div(dom_vec)
```

Comment:
That's a test of the sort "runs at all", which is valid in a `test_<class>_init()` type test, just for functions. 

Speaking of which, I think the test functions do a bit too much at a time. In theory, each test function should test exactly one scenario and one parameter. This is probably a bit too fine-grained, but having small and isolated tests is a big plus when you need to pin down an error source. I suggest that you split the tests at least into 1D and nD and also test different schemes separately. It may be worth making a fixture for that.